src/main/java/com/shaft/tools/io/YAMLFileManager.java

```diff
@@ -0,0 +1,404 @@
+     * @param key the key that contains one or more index
+     * @return a {@link List<Integer>} that contains one or more index
+     */
+    private List<Integer> parseKeyList(String key) {
+        var indexes = new ArrayList<Integer>();
+
+        var matcher = Pattern
+                .compile(NUMBER_IN_SQUARE_BRACKETS_REGEX).matcher(key);
+
+        while (matcher.find()) {
+            indexes.add(Integer.parseInt(
+                    matcher
+                            .group()
+                            .replaceAll(SQUARE_BRACKETS_REGEX, "")
+            ));
+        }
+
+        return indexes;
+    }
+
+    /**
+     * Used internally to parse {@link Object} to be any type
+     *
+     * @param obj the value wanted to be parsed
+     * @param clazz the class that data wanted to be parsed for
+     * @return the value as {@link T}
+     * @param <T> the wanted type
+     */
+    private <T> T parseObjectTo(Object obj, Class<T> clazz) {
+        T v;
+        try {
+            v = clazz.cast(obj);
+        } catch (ClassCastException rootCauseException) {
+            var errorMsg = "Can't parse the value of [" + obj + "] to be of type [" + clazz.getSimpleName() + "]";
+            sendFailureCompo(errorMsg, rootCauseException);
+            throw rootCauseException;
+        }
+        return v;
+    }
+
+    /**
+     * Used internally to parse {@link Object} to be a {@link List<T>}
+     *
+     * @param obj the value wanted to be parsed
+     * @param clazz the class that data wanted to be parsed for
+     * @return the value as {@link List<T>}
+     * @param <T> the wanted type
+     */
+    private <T> List<T> parseObjectToList(Object obj, Class<T> clazz) {
+        if (obj instanceof List<?> list) {
+            return list.stream()
+                    .map(item -> parseObjectTo(item, clazz)).toList();
+        }
+
+        var errorMsg = "Can't parse the value of [" + obj + "] to be list";
+        sendFailureCompo(errorMsg);
+        throw new RuntimeException(errorMsg);
+    }
+
+    /**
+     * Used internally to parse {@link Object} to be a {@link Map} of {@link String} and {@link T}
+     *
+     * @param obj the value wanted to be parsed
+     * @param clazz the class that data wanted to be parsed for
+     * @return the value as {@link Map} of {@link String} and {@link T}
+     * @param <T> the wanted type
+     */
+    private <T> Map<String, T> parseObjectToMap(Object obj, Class<T> clazz) {
+        if (obj instanceof Map<?, ?> map) {
+            var nMap = new HashMap<String, T>();
+            map.forEach(
+                    (k, v) -> nMap.put(k.toString(), parseObjectTo(v, clazz))
+            );
+            return nMap;
+        }
+        var errorMsg = "Can't parse the value of [" + obj + "] to be map";
+        sendFailureCompo(errorMsg);
+        throw new RuntimeException(errorMsg);
+    }
+
+    private void sendFailureCompo(String msg, Exception... rootCauseException) {
```

Comment:
We can use this common method to apply the failure pattern mentioned in the previous review comment, and ensure that all caught exceptions and failed methods call it.